polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp

```diff
@@ -491,78 +492,73 @@ void MemorySelector::analyze(LoopLikeOpInterface loop, AccessKind accessKind) {
       continue;
     }
 
-    // Get the inter-thread access pattern and classify the memory access.
+    memRefAccessToMemSpace[memRef] =
+        selectMemorySpace(*memAccess, threadVars, accessKind);
+
+    LLVM_DEBUG({
+      if (memRefAccessToMemSpace.at(memRef) == MemorySpace::Shared)
+        llvm::dbgs().indent(2) << "shared memory space\n";
+      else {
+        assert(memRefAccessToMemSpace.at(memRef) == MemorySpace::Global);
+        llvm::dbgs().indent(2) << "global memory space\n";
+      }
+    });
+  }
+}
+
+MemorySelector::MemorySpace
+MemorySelector::selectMemorySpace(const MemoryAccess &memAccess,
+                                  ArrayRef<Value> threadVars,
+                                  AccessKind accessKind) const {
+  const unsigned gridDimension = threadVars.size();
+
+  // Whether the memory access is (partially) coalesced or not.
+  auto isCoalesced = [&](const MemoryAccess &memAccess) {
     MemoryAccessMatrix interThreadMatrix =
-        memAccess->getInterThreadAccessMatrix(gridDim);
+        memAccess.getInterThreadAccessMatrix(gridDimension);
     MemoryAccessPattern interThreadAccessPattern =
-        MemoryAccess::classify(interThreadMatrix, memAccess->getOffsetVector());
+        MemoryAccess::classify(interThreadMatrix, memAccess.getOffsetVector());
 
     switch (interThreadAccessPattern) {
     case Linear:
     case Reverse:
     case ReverseLinear:
       // These patterns imply fully coalesced memory accesses.
-      memRefAccessToMemSpace[memRef] = MemorySpace::Global;
-      break;
+      return true;
     case Shifted:
     case LinearShifted:
     case ReverseLinearShifted:
     case LinearOverlapped:
     case ReverseLinearOverlapped:
       // These patterns imply partially coalesced memory accesses.
-      memRefAccessToMemSpace[memRef] = MemorySpace::Global;
-      break;
-    case Strided:
```

Comment:
Changed the algorithm used to select shared memory to match the one in the paper.